### DISCUSSION TITLE
Sample vs simulate behaviour

### DISCUSSION THREAD

**Discussion by ra1u** [clash-lang/clash-compiler#2029] [Q&A] [answered]
Is this bahaviour expected (notice difference between test1 and test2 output )?

Can I use sample to behave as simulate and that output does not start with 2 zeros?

``` haskell
import Clash.Prelude
import qualified Prelude as P

test1 = P.take 10 $ sample
                   $ register @System (0 :: (Signed 5))
                   $ fromList
                   $ P.repeat 1
-- test1 == [0,0,1,1,1,1,1,1,1,1]

test2 = P.take 10 $ simulate
                  ( register @System (0 :: (Signed 5)))
                  $ P.repeat 1
-- test2 == [0,1,1,1,1,1,1,1,1,1]
```

My version is 1.4.2.

**Comment by martijnbastiaan** [comment, marked as answer]
> Is this bahaviour expected (notice difference between test1 and test2 output )?

Yes, insofar as that the documentation of sample says:

> If the given component has not yet been given a clock, reset, or enable line, sample will supply them. The reset will be asserted for a single cycle. **sample will not drop the value produced by the circuit while the reset was asserted**. If you want this, or if you want more than a single cycle reset, consider using sampleWithReset.

Documentation of simulate:

> Where 'System' denotes the /domain/ to simulate on. The reset line is asserted for a single cycle. The first value is therefore supplied twice to the circuit: once while reset is high, and once directly after. **The first /output/ value (the value produced while the reset is asserted) is dropped**.

(It does look like the Haddock is missing this simulate documentation for some reason.)

> Is this bahaviour expected (notice difference between test1 and test2 output )?

No, insofar as that you didn't expect it, and I think it's a bit of a wart to have them behave differently.

> Can I use sample to behave as simulate and that output does not start with 2 zeros?

Yes, use `sampleWithReset`.
